AsyncDisplayKit/Details/ASCollectionDataController.mm

```diff
@@ -150,6 +150,55 @@ - (void)willMoveSection:(NSInteger)section toSection:(NSInteger)newSection
   }
 }
 
+- (void)prepareForInsertRowsAtIndexPaths:(NSArray<NSIndexPath *> *)indexPaths
+{
+  for (NSString *kind in [self supplementaryKinds]) {
+    LOG(@"Populating elements of kind: %@, for index paths: %@", kind, indexPaths);
+    NSMutableArray<ASIndexedNodeContext *> *contexts = [NSMutableArray array];
+    [self _populateSupplementaryNodesOfKind:kind atIndexPaths:indexPaths mutableContexts:contexts];
+    _pendingContexts[kind] = contexts;
+  }
+}
+
+- (void)willInsertRowsAtIndexPaths:(NSArray<NSIndexPath *> *)indexPaths
+{
+  [_pendingContexts enumerateKeysAndObjectsUsingBlock:^(NSString *kind, NSMutableArray<ASIndexedNodeContext *> *contexts, BOOL *stop) {
+    [self batchLayoutNodesFromContexts:contexts ofKind:kind completion:^(NSArray<ASCellNode *> *nodes, NSArray<NSIndexPath *> *indexPaths) {
+      [self insertNodes:nodes ofKind:kind atIndexPaths:indexPaths completion:nil];
+    }];
+    [_pendingContexts removeObjectForKey:kind];
```

Comment:
It seems like it doesn't throw an exception but does not do what you expect either. Try this in a playground: 
```
var mutableArray = NSMutableArray(array: [1,2,3,4,5,6])

mutableArray.enumerateObjectsUsingBlock { (object, idx, stop) in
    print("\(idx): \(object)")
    mutableArray.removeObject(object)
}
```